zonemaster/zonemaster-engine
# Fix code in Basic01

## Purpose

This PR fixes the implementation of Basic01 by adding missing code due to an oversight during the initial implementation update (#1212).

It also adds conditional checks to the start of the Basic module and Test Cases (see ["Scope" section of Basic01](https://github.com/zonemaster/zonemaster/blob/v2023.1/docs/public/specifications/tests/Basic-TP/basic01.md?plain=1#L41-L44)) that were also missed.
## Context

Fixes #1248 

## Changes

- Update and add missing code (step 7.3 of the specification)
- Add missing conditional checks for the start of the Basic module and test cases ("scope" section of the specification)
- Refactoring
- Update unitary tests

## How to test this PR

Check that this zone (from #1248) works correctly. Note that we do not have an unitary test for that kind of zone configuration yet.
```
$ zonemaster-cli zm2.zm1.andreasschulze.de --test=basic/basic01 --level=INFO
Seconds Level    Message
======= ======== =======
   0.01 INFO     Using version v4.7.0 of the Zonemaster engine.
   4.22 INFO     The zone "zm2.zm1.andreasschulze.de" is found.
   4.22 INFO     The parent zone is "andreasschulze.de." as returned from name servers "dahlem.somaf.de/185.183.157.243;dahlem.somaf.de/2a03:4000:1d:8e6::1;dallas.signing-milter.org/130.255.77.178;dallas.signing-milter.org/2a02:e00:ffec:8cd::1".
```

Check that a non existing zone, normal test makes Zonemaster:
 - Skip Basic02
 - Run Basic03
 - Abort the testing suite
```
$ zonemaster-cli example.fake --level=DEBUG --show-testcase --raw | grep TEST_CASE_START
   0.00 DEBUG    BASIC00        TEST_CASE_START   testcase=basic00
   0.00 DEBUG    BASIC01        TEST_CASE_START   testcase=basic01
   3.35 DEBUG    BASIC03        TEST_CASE_START   testcase=basic03
```

Check that a non existing zone, undelegated test makes Zonemaster:
 - Run Basic02
 - Run Basic03
 - Continue the testing suite
```
$ zonemaster-cli example.fake --level=DEBUG --show-testcase --raw --ns=ns1.example.fake/192.168.1.100 | grep TEST_CASE_START
   0.00 DEBUG    BASIC00        TEST_CASE_START   testcase=basic00
   0.00 DEBUG    BASIC01        TEST_CASE_START   testcase=basic01
   1.90 DEBUG    BASIC02        TEST_CASE_START   testcase=basic02
  11.93 DEBUG    BASIC03        TEST_CASE_START   testcase=basic03
  11.93 DEBUG    ADDRESS01      TEST_CASE_START   testcase=address01
  11.93 DEBUG    ADDRESS02      TEST_CASE_START   testcase=address02
  12.91 DEBUG    CONNECTIVITY01 TEST_CASE_START   testcase=connectivity01
  12.91 DEBUG    CONNECTIVITY02 TEST_CASE_START   testcase=connectivity02
  [...]
```
